---
title: "Training CD cohort GSE16879"
author: "Reshetnikova G"
date: "2025-01-16"
output: html_document
---
Загрузка библиотек:
```{r}
source("req.R") # if knitr is not installed, run this string in consol
devtools::install_github("kevinblighe/CorLevelPlot")
```

```{r setup, include=FALSE}
main_dir <- dirname(rstudioapi::getSourceEditorContext()$path) 
setwd(main_dir)
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(GEOquery)
library(tidyverse)
library(tibble)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(affy)
library(GEOquery)
library(limma)
library(umap)
library(affy)
library(vegan)
library(pheatmap)
library(openxlsx)
library(clusterProfiler)
library(org.Hs.eg.db)
library(AnnotationDbi)
library(gage)
library(ggrepel)
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)
library(DOSE)
library(impute)
library(WGCNA)
library(DESeq2)
library(GEOquery)
library(tidyverse)
library(CorLevelPlot)
library(gridExtra)
library(readr)
library(VennDiagram)
main_dir <- dirname(rstudioapi::getSourceEditorContext()$path) 
setwd(main_dir)
```
# Обзор датасета

Info:
Mucosal biopsies were obtained at endoscopy in actively inflamed mucosa from 61 IBD patients:
- 24 ulcerative colitis (UC), 
- 19 Crohn’s colitis (CDc), 
- 18 Crohn’s ileitis (CDi)), 
refractory to corticosteroids and/or immunosuppression, before and 4-6 weeks after (except for 1 CDc patient) their first infliximab infusion and in normal mucosa from 12 control patients (6 colon and 6 ileum). The patients were classified for response to infliximab based on endoscopic and histologic findings at 4-6 weeks after first infliximab treatment. Total RNA was isolated from intestinal mucosal biopsies, labelled and hybridized to Affymetrix Human Genome U133 Plus 2.0 Arrays.
```{r}
# Download the data
gse_16879 <- getGEO("GSE16879")
```
## Metadata
Load the data
```{r}
# get metadata
metadata <- pData(phenoData(gse_16879[[1]]))
```

Save useful columns
```{r}
# edit table
metadata <- metadata %>% dplyr::select(geo_accession, title, source_name_ch1, `tissue:ch1`, `response to infliximab:ch1`, `disease:ch1`, `before or after first infliximab treatment:ch1`)

```

Rename columns

```{r}
metadata <- metadata %>%
  dplyr::rename(
    Geo_accession = geo_accession,
    Title = title,
    Sourse_Name = `source_name_ch1`,
    Tissue = `tissue:ch1`,
    Response = `response to infliximab:ch1`,
    Disease = `disease:ch1`,
    Timeline = `before or after first infliximab treatment:ch1`
  )

```

Save the metadata about groups that should be compared
Сохраню отдельно метаданные о группах, которые буду сравнивать
```{r}
# 1. All before treatment
meta_all_beforeT <- metadata %>%
  filter(grepl("beforeT", Title))

# 2. All after treatment
meta_all_afterT <- metadata %>%
  filter(grepl("afterT", Title))

# 3. Responders before treatment
meta_all_R_beforeT <- metadata %>%
  filter(grepl("(?<!N)R[0-9]*_beforeT", Title, perl = TRUE))

# 4. Responders after treatment
meta_all_R_afterT <- metadata %>%
  filter(grepl("(?<!N)R[0-9]*_afterT", Title, perl = TRUE))
# для CDcR2 есть информация только ДО лечения, а после -- нет.

# 5. Non-responders before treatment
meta_all_NR_beforeT <- metadata %>%
  filter(grepl("NR[0-9]*_beforeT", Title, perl = TRUE))

# 6. Npn-responders after treatment
meta_all_NR_afterT <- metadata %>%
  filter(grepl("NR[0-9]*_afterT", Title, perl = TRUE))

```

## Normalize expression data

Download expression data and normalize it

```{r}
# get supplementary files
getGEOSuppFiles("GSE16879", baseDir = '../../../data/CD/train')

# If the download fails with an error, then you should download the data directly to the data/CD/train folder using the link (https://ftp.ncbi.nlm.nih.gov/geo/series/GSE16nnn/GSE16879/suppl//GSE16879_RAW.tar?tool=geoquery ), create a folder GSE16879 and put the downloaded archive with data there (GSE16879_RAW.tar)
```

```{r}
# untar files
untar("../../../data/CD/train/GSE16879/GSE16879_RAW.tar", exdir = '../../../data/CD/train/GSE16879/GSE16879_data')

# reading in .cel files
raw.data <- ReadAffy(celfile.path = "../../../data/CD/train/GSE16879/GSE16879_data")

# performing RMA normalization
normalized.data <- rma(raw.data) 

# get expression estimates
normalized.expr <- as.data.frame(exprs(normalized.data))

colnames(normalized.expr) <- gsub("\\.CEL\\.gz$", "", colnames(normalized.expr))

```


Now we will add the Gene.Symbol to each sample from normalized.expr.

```{r}
# map probe IDs to gene symbols
gse_gs <- getGEO("GSE16879", GSEMatrix = TRUE)

# fetch feature data to get ID - gene symbol mapping
feature.data <- gse_gs$GSE16879_series_matrix.txt.gz@featureData@data

# subset
feature.data <- feature.data[, c(1,11)]

# get ID column from normalized.expr
normalized.expr <- normalized.expr %>%
  rownames_to_column(var = "ID")

# merge DFs
gene_normalized.expr <- inner_join(normalized.expr, feature.data, by = "ID")

```


Gaps from gene_normalized.expr should be deleted:

```{r}
# Replace Gene Symbol to Gene.Symbol

gene_normalized.expr <- gene_normalized.expr %>%
  dplyr::rename(Gene.Symbol = `Gene Symbol`)

filtered_gene_normalized.expr <- gene_normalized.expr %>%
  filter(Gene.Symbol != "" & !is.na(Gene.Symbol))

nrow(gene_normalized.expr)-nrow(filtered_gene_normalized.expr)

```
 8893 strings without Gene.Symbol were deleted 


Some samples correspond to the same genes, so we will find duplicate genes and average the values
```{r}
# We group the data by Gene Symbol and calculate the average expression value
mean_norm.expr <- filtered_gene_normalized.expr %>%
  group_by(`Gene.Symbol`) %>%
  summarise(across(where(is.numeric), \(x) mean(x, na.rm = TRUE)))

write.table(mean_norm.expr, "../../../tables/CD/train/mean_norm.expr.tsv", sep = "\t", row.names = TRUE, quote = FALSE)
```

# PCA
Now, having averaged expression values, we can perform PCA.

```{r}
mean_norm.expr <- read.table("../../../tables/CD/train/mean_norm.expr.tsv", sep = "\t", header=TRUE)

# For PCA, there should be a matrix without rows, so let's filter out the names of the genes.
pca_matrix <- mean_norm.expr %>%
  dplyr::select(-Gene.Symbol) %>%
  as.matrix()

# Save gene names separately
gene_symbols <- mean_norm.expr$Gene.Symbol

# Transpose matrix
pca_matrix <- t(pca_matrix)

# In our Univrsity we usually preformed РСА with vegan library


# Matrix preaparation
pca_matrix_scaled <- decostand(pca_matrix, method = "standardize")

# PCA
pca_vegan <- rda(pca_matrix_scaled, scale = TRUE)

# summary(pca_vegan) -- is too overloaded, Scree is better to evaluate components contribution

screeplot(pca_vegan, main = "Scree Plot: PCA on Gene Expression")
```
We can see that first 4 PC make remarkable contribution. Further we will work with РС1 and РС2, but РС3 и РС4 are alse will be saved to find there other info if needed

```{r}
# Download the values of the first 4 components
pca_scores <- scores(pca_vegan, display = "sites", choices = 1:4)
pca_scores_df <- as.data.frame(pca_scores)

# convert rownames into column with ID
pca_scores_df <- pca_scores_df %>%
  rownames_to_column(var = "ID")

# Merge PCs and metadata by ID
pca_df <- dplyr::left_join(pca_scores_df, metadata, by = c("ID" = "Geo_accession"))

```
We will be comparing groups, so we will build tables based on comparison groups.

```{r}
# The first comparison group: EVERYTHING before and after treatment. Combine the all_beforeT and all_afterT tables
meta_all_beforeT_and_afterT <- bind_rows(meta_all_beforeT, meta_all_afterT)
pca_all_beforeT_and_afterT <- inner_join(pca_scores_df, meta_all_beforeT_and_afterT, by = c("ID" = "Geo_accession"))

# The second comparison group: ALL respondents before and after treatment
meta_R_beforeT_and_afterT <- bind_rows(meta_all_R_beforeT, meta_all_R_afterT)
pca_R_beforeT_and_afterT <- inner_join(pca_scores_df, meta_R_beforeT_and_afterT, by = c("ID" = "Geo_accession"))

# The third comparison group: ALL nonresponders before and after treatment
meta_NR_beforeT_and_afterT <- bind_rows(meta_all_NR_beforeT, meta_all_NR_afterT)
pca_NR_beforeT_and_afterT <- inner_join(pca_scores_df, meta_NR_beforeT_and_afterT, by = c("ID" = "Geo_accession"))

# The fourth comparison group: respondents BEFORE versus non-respondents BEFORE
meta_RNR_beforeT <- bind_rows(meta_all_R_beforeT, meta_all_NR_beforeT)
pca_RNR_beforeT <- inner_join(pca_scores_df, meta_RNR_beforeT, by = c("ID" = "Geo_accession"))

# The fifth comparison group: POST-responders versus non-responders AFTER
meta_RNR_afterT <- bind_rows(meta_all_R_afterT, meta_all_NR_afterT)
pca_RNR_afterT <- inner_join(pca_scores_df, meta_RNR_afterT, by = c("ID" = "Geo_accession"))
```

## All patients BEFORE and AFTER

1. All groups BEFORE and AFTER treatment
```{r}
# PCA graph for the first group
pca_1 <- ggplot(pca_all_beforeT_and_afterT, aes(x = PC1, y = PC2, color = Timeline)) +
  geom_point(size = 4, alpha = 0.8) +
  labs(title = "PCA of IBD Patients (PC1 vs PC2)", x = "PC1", y = "PC2") +
  theme_minimal() +
  scale_color_manual(values = c("red", "blue")) 

```
## Respondents BEFORE and AFTER
2. Respondents BEFORE and respondents AFTER
```{r}
# PCA graph for the second group
pca_2 <- ggplot(pca_R_beforeT_and_afterT, aes(x = PC1, y = PC2, color = Timeline)) +
  geom_point(size = 4, alpha = 0.8) +
  labs(title = "PCA of IBD Patients (PC1 vs PC2)", x = "PC1", y = "PC2") +
  theme_minimal() +
  scale_color_manual(values = c("red", "blue")) 
```
## Non-responders BEFORE and AFTER
3. Nonresponders BEFORE and nonresponders AFTER
```{r}
# PCA graph for the third group
pca_3 <- ggplot(pca_NR_beforeT_and_afterT, aes(x = PC1, y = PC2, color = Timeline)) +
  geom_point(size = 4, alpha = 0.8) +
  labs(title = "PCA of IBD Patients (PC1 vs PC2)", x = "PC1", y = "PC2") +
  theme_minimal() +
  scale_color_manual(values = c("red", "blue")) 
```
## Responders and Non-Responders BEFORE
4. Respondents BEFORE and non-respondents BEFORE
```{r}
# PCA graph for the fourth group
pca_4 <- ggplot(pca_RNR_beforeT, aes(x = PC1, y = PC2, color = Response)) +
  geom_point(size = 4, alpha = 0.8) +
  labs(title = "PCA of IBD Patients (PC1 vs PC2)", x = "PC1", y = "PC2") +
  theme_minimal() +
  scale_color_manual(values = c("red", "blue")) 
```
## Responders and Non-Responders AFTER
5. POST-responders and non-responders AFTER
```{r}
# PCA graph for the fifth group
pca_5 <- ggplot(pca_RNR_afterT, aes(x = PC1, y = PC2, color = Response)) +
  geom_point(size = 4, alpha = 0.8) +
  labs(title = "PCA of IBD Patients (PC1 vs PC2)", x = "PC1", y = "PC2") +
  theme_minimal() +
  scale_color_manual(values = c("red", "blue"))
```

## All types of diseases (UC, Cdi, CDc, Control)
```{r}
pca_df$Disease <- as.factor(pca_df$Disease)
pca_df$Response <- as.factor(pca_df$Response)


# Make column `Disease_Updated`
pca_df <- pca_df %>%
  dplyr::mutate(All_Types_Of_Diseases = case_when(
    Disease == "CD" & Tissue == "Ileum" ~ "CDi",
    Disease == "CD" & Tissue == "Colon" ~ "CDc",
    TRUE ~ Disease
  ))

# Check what we obtained
table(pca_df$All_Types_Of_Diseases)

# PCA graph with colors by diagnosis
pca_6 <- ggplot(pca_df, aes(x = PC1, y = PC2, color = All_Types_Of_Diseases)) +
  geom_point(size = 4, alpha = 0.8) +
  labs(title = "PCA of IBD Patients (PC1 vs PC2)", x = "PC1", y = "PC2") +
  theme_minimal() +
  scale_color_manual(values = c("red", "blue", "green", "black")) 

```
## All patients by response (Yes, No, Control)
Or by response
```{r}
# PCA-graph with colors based on response
pca_7 <- ggplot(pca_df, aes(x = PC1, y = PC2, color = Response)) +
  geom_point(size = 4, alpha = 0.8) +
  labs(title = "PCA of IBD Patients (PC1 vs PC2)", x = "PC1", y = "PC2") +
  theme_minimal() +
  scale_color_manual(values = c("red", "blue", "green", "black")) 
```

## Save PCA graphs
```{r}
dir.create("../../../images/CD/train/PCA", recursive = TRUE, showWarnings = FALSE) 

pca_list <- list(pca_1, pca_2, pca_3, pca_4, pca_5, pca_6, pca_7)

file_names <- c(
  "1_All_Before_After.svg",  # Все пациенты ДО и ПОСЛЕ
  "2_R_Before_After.svg",    # Респондеры ДО и ПОСЛЕ
  "3_NR_Before_After.svg",   # Нереспондеры ДО и ПОСЛЕ
  "4_R_NR_Before.svg",       # Респондеры и Нереспондеры ДО
  "5_R_NR_After.svg",        # Респондеры и Нереспондеры ПОСЛЕ
  "6_Disease_Types.svg",     # Все типы болезней (UC, Cdi, CDc, Control)
  "7_Response_Status.svg"    # Все пациенты по респонсу (Yes, No, Control)
)

for (i in seq_along(pca_list)) {
  file_path <- paste0("../../../images/CD/train/PCA/", file_names[i])
  svg(file_path, width = 9, height = 6)
  print(pca_list[[i]])
  dev.off()
}

```

# Differential expression analysis

## All patients

Converting metadata for more convenient operation
```{r}
# convert the before-after to metadata:
# 1. Convert the levels of the "Timeline" factor
meta_all_beforeT_and_afterT$Timeline <- factor(meta_all_beforeT_and_afterT$Timeline, 
                                          levels = c("Before first infliximab treatment", 
                                                     "After first infliximab treatment"),
                                          labels = c("before", "after"))


# 2. Converting the levels of the "Timeline" factor
meta_R_beforeT_and_afterT$Timeline <- factor(meta_R_beforeT_and_afterT$Timeline, 
                                          levels = c("Before first infliximab treatment", 
                                                     "After first infliximab treatment"),
                                          labels = c("before", "after"))


# 3. Converting the levels of the "Timeline" factor
meta_NR_beforeT_and_afterT$Timeline <- factor(meta_NR_beforeT_and_afterT$Timeline, 
                                          levels = c("Before first infliximab treatment", 
                                                     "After first infliximab treatment"),
                                          labels = c("before", "after"))

# 4. Converting the levels of the "Timeline" factor
meta_RNR_beforeT$Timeline <- factor(meta_RNR_beforeT$Timeline, 
                                          levels = c("Before first infliximab treatment", 
                                                     "After first infliximab treatment"),
                                          labels = c("before", "after"))


# 5. Converting the levels of the "Timeline" factor
meta_RNR_afterT$Timeline <- factor(meta_RNR_afterT$Timeline, 
                                          levels = c("Before first infliximab treatment", 
                                                     "After first infliximab treatment"),
                                          labels = c("before", "after"))


# For groups 4 and 5, it was not necessary to convert the timeline to the before-after factor at all, because in them it is interesting for us to compare respondents and non-respondents.
# Convert to the Response factor in groups 4 and 5

meta_RNR_beforeT$Response <- factor(meta_RNR_beforeT$Response, levels = c("No", "Yes"))
meta_RNR_afterT$Response <- factor(meta_RNR_afterT$Response, levels = c("No", "Yes"))


```
We will store the expression data separately according to the metadata for each of the groups.:

1. The first comparison group: EVERYTHING before and after treatment. all_beforeT_and_afterT
2. The second comparison group: ALL respondents before and after treatment. R_beforeT_and_afterT
3. The third comparison group: ALL nonresponders before and after treatment. NR_beforeT_and_afterT
4. The fourth comparison group: nonresponders BEFORE versus respondents BEFORE. RNR_beforeT
5. The fifth comparison group: nonresponders AFTER versus respondents AFTER. RNR_afterT


```{r}
# mean_norm.expr stores annotated information about the gene expression of each sample. To compare groups, we will compare the ID from the metadata of the groups (meta_all_...) with the ID from mean_norm.expr.
# Convert it to data.frame (so that rownames can be set)
mean_norm.expr <- as.data.frame(mean_norm.expr)
# Make genes as rownames, and delete the column
rownames(mean_norm.expr) <- mean_norm.expr$Gene.Symbol 
mean_norm.expr <- mean_norm.expr[, !colnames(mean_norm.expr) %in% "Gene.Symbol"]


expr_1 <- mean_norm.expr[, meta_all_beforeT_and_afterT$Geo_accession]
expr_2 <- mean_norm.expr[, meta_R_beforeT_and_afterT$Geo_accession]
expr_3 <- mean_norm.expr[, meta_NR_beforeT_and_afterT$Geo_accession]
expr_4 <- mean_norm.expr[, meta_RNR_beforeT$Geo_accession]
expr_5 <- mean_norm.expr[, meta_RNR_afterT$Geo_accession]

```

As I wrote above, we are interested in groups 1-3 in order to find the difference between groups of patients before and after treatment, and groups 4-5 - to see
the difference between responders and non-responders. Of course, this will also affect the construction of the experiment matrix. Therefore, first we will construct matrices for groups 1-3 in order to search for TAGs in various before-after groups.

### Design matrix 1-3
```{r}
# Groups 1-3. Before and after

design_1 <- model.matrix(~0 + meta_all_beforeT_and_afterT$Timeline)
colnames(design_1) <- levels(meta_all_beforeT_and_afterT$Timeline)

design_2 <- model.matrix(~0 + meta_R_beforeT_and_afterT$Timeline)
colnames(design_2) <- levels(meta_R_beforeT_and_afterT$Timeline)

design_3 <- model.matrix(~0 + meta_NR_beforeT_and_afterT$Timeline)
colnames(design_3) <- levels(meta_NR_beforeT_and_afterT$Timeline)

# As far as I understand, ~0 allows us to create models without an intersept, and each group becomes a separate variable. 
```

### Design matrix 4-5
Then we will build matrices for groups 4-5 to search for TAGs in the responder-nonresponder groups.

```{r}
# Groups 4-5. Non-responders
design_4 <- model.matrix(~0 + meta_RNR_beforeT$Response)
colnames(design_4) <- levels(meta_RNR_beforeT$Response)

design_5 <- model.matrix(~0 + meta_RNR_afterT$Response)
colnames(design_5) <- levels(meta_RNR_afterT$Response)

```
makeContrasts. As far as I understand, we have defined groups in the design matrix, and makeContrast allows us to compare these groups.:

### Contrast matrix 1-3
```{r}
# Groups 1-3

 Group 1: All samples before and after treatment
contrast_1 <- makeContrasts(before - after, levels = design_1)

# Группа 2: Only responders before and after
contrast_2 <- makeContrasts(before - after, levels = design_2)

# Группа 3: Only non-responders before and after
contrast_3 <- makeContrasts(before - after, levels = design_3)

```
### Contrast matrix 4-5
```{r}
# groups 4-5

# Group 4: non-responders vs responders BEFORE treatment
contrast_4 <- makeContrasts(No - Yes, levels = design_4)

# Group 5: non-responders vs responders AFTER treatment
contrast_5 <- makeContrasts(No - Yes, levels = design_5)

```

Build a model for each comparison group:
```{r}
fit_1 <- lmFit(expr_1, design_1)
fit_2 <- lmFit(expr_2, design_2)
fit_3 <- lmFit(expr_3, design_3)

fit_4 <- lmFit(expr_4, design_4)
fit_5 <- lmFit(expr_5, design_5)
```
These linear models were built on the basis of a design matrix, but without comparing groups directly, so we use contrasts.fit() to "supplement" the resulting models with information about comparing groups (contrasting matrices):

```{r}
fit2_1 <- contrasts.fit(fit_1, contrast_1)
fit2_2 <- contrasts.fit(fit_2, contrast_2)
fit2_3 <- contrasts.fit(fit_3, contrast_3)

fit2_4 <- contrasts.fit(fit_4, contrast_4)
fit2_5 <- contrasts.fit(fit_5, contrast_5)
```

We adjust the statistics and get the p-value values and others using eBayes

```{r}
fit2_1 <- eBayes(fit2_1)
fit2_2 <- eBayes(fit2_2)
fit2_3 <- eBayes(fit2_3)

fit2_4 <- eBayes(fit2_4)
fit2_5 <- eBayes(fit2_5)
```

### Results for all genes
We have quite a lot of data, so we will use the Benjamin-Hochberg correction to identify DEGs. In addition, we will save the files with the results separately.
```{r}

dir.create("../../../tables/CD/train/DEGs_All_Results", recursive = TRUE, showWarnings = FALSE) 

DEG_results_1 <- topTable(fit2_1, coef = 1, adjust = "BH", number = Inf)
write.table(DEG_results_1, "../../../tables/CD/train/DEGs_All_Results/DEG_results_1.txt", sep = "\t", quote = FALSE, row.names = TRUE)

DEG_results_2 <- topTable(fit2_2, coef = 1, adjust = "BH", number = Inf)
write.table(DEG_results_2, "../../../tables/CD/train/DEGs_All_Results/DEG_results_2.txt", sep = "\t", quote = FALSE, row.names = TRUE)

DEG_results_3 <- topTable(fit2_3, coef = 1, adjust = "BH", number = Inf)
write.table(DEG_results_3, "../../../tables/CD/train/DEGs_All_Results/DEG_results_3.txt", sep = "\t", quote = FALSE, row.names = TRUE)


DEG_results_4 <- topTable(fit2_4, coef = 1, adjust = "BH", number = Inf)
write.table(DEG_results_4, "../../../tables/CD/train/DEGs_All_Results/DEG_results_4.txt", sep = "\t", quote = FALSE, row.names = TRUE)

DEG_results_5 <- topTable(fit2_5, coef = 1, adjust = "BH", number = Inf)
write.table(DEG_results_5, "../../../tables/CD/train/DEGs_All_Results/DEG_results_5.txt", sep = "\t", quote = FALSE, row.names = TRUE)


```

### DEGs
Let's filter out the significant genes and save the results:

```{r}
# Filtering significant DEGs (p.adj < 0.05, |logFC| > 1) and saving as .tsv
deg_sign_1 <- subset(DEG_results_1, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(deg_sign_1, "../../../tables/CD/train/DEGs_All_Results/deg_sign_1.tsv", sep = "\t", quote = FALSE, row.names = TRUE, col.names = NA)

deg_sign_2 <- subset(DEG_results_2, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(deg_sign_2, "../../../tables/CD/train/DEGs_All_Results/deg_sign_2.tsv", sep = "\t", quote = FALSE, row.names = TRUE, col.names = NA)

deg_sign_3 <- subset(DEG_results_3, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(deg_sign_3, "../../../tables/CD/train/DEGs_All_Results/deg_sign_3.tsv", sep = "\t", quote = FALSE, row.names = TRUE, col.names = NA)

deg_sign_4 <- subset(DEG_results_4, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(deg_sign_4, "../../../tables/CD/train/DEGs_All_Results/deg_sign_4.tsv", sep = "\t", quote = FALSE, row.names = TRUE, col.names = NA)

deg_sign_5 <- subset(DEG_results_5, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(deg_sign_5, "../../../tables/CD/train/DEGs_All_Results/deg_sign_5.tsv", sep = "\t", quote = FALSE, row.names = TRUE, col.names = NA)


```


### Statistics
Statistics on adj.P.Val only. Let's calculate how many DEGs adj.P.Val < 0.05 have

```{r}
deg_stats_padj <- data.frame(
  Группа = c("Группа 1", "Группа 2", "Группа 3", "Группа 4", "Группа 5"),
  Всего_DEG = c(nrow(DEG_results_1), nrow(DEG_results_2), nrow(DEG_results_3), nrow(DEG_results_4), nrow(DEG_results_5)),
  Значимые_DEG = c(sum(DEG_results_1$adj.P.Val < 0.05),
                   sum(DEG_results_2$adj.P.Val < 0.05),
                   sum(DEG_results_3$adj.P.Val < 0.05),
                   sum(DEG_results_4$adj.P.Val < 0.05),
                   sum(DEG_results_5$adj.P.Val < 0.05)),
  Процент_значимых = round(c(
    (sum(DEG_results_1$adj.P.Val < 0.05) / nrow(DEG_results_1)) * 100,
    (sum(DEG_results_2$adj.P.Val < 0.05) / nrow(DEG_results_2)) * 100,
    (sum(DEG_results_3$adj.P.Val < 0.05) / nrow(DEG_results_3)) * 100,
    (sum(DEG_results_4$adj.P.Val < 0.05) / nrow(DEG_results_4)) * 100,
    (sum(DEG_results_5$adj.P.Val < 0.05) / nrow(DEG_results_5)) * 100
  ), 2)
)

print(deg_stats_padj)


```
Interestingly, in the third group of DEGs with adj.P.Val < 0.05 -- 0

```{r}
deg_stats_padj_lfc <- data.frame(
  Группа = c("Группа 1", "Группа 2", "Группа 3", "Группа 4", "Группа 5"),
  Всего_DEG = c(nrow(DEG_results_1), nrow(DEG_results_2), nrow(DEG_results_3), nrow(DEG_results_4), nrow(DEG_results_5)),
  Значимые_DEG = c(sum(DEG_results_1$adj.P.Val < 0.05 & abs(DEG_results_1$logFC) > 1),
                   sum(DEG_results_2$adj.P.Val < 0.05 & abs(DEG_results_2$logFC) > 1),
                   sum(DEG_results_3$adj.P.Val < 0.05 & abs(DEG_results_3$logFC) > 1),
                   sum(DEG_results_4$adj.P.Val < 0.05 & abs(DEG_results_4$logFC) > 1),
                   sum(DEG_results_5$adj.P.Val < 0.05 & abs(DEG_results_5$logFC) > 1)),
  Процент_значимых = round(c(
    (sum(DEG_results_1$adj.P.Val < 0.05 & abs(DEG_results_1$logFC) > 1) / nrow(DEG_results_1)) * 100,
    (sum(DEG_results_2$adj.P.Val < 0.05 & abs(DEG_results_2$logFC) > 1) / nrow(DEG_results_2)) * 100,
    (sum(DEG_results_3$adj.P.Val < 0.05 & abs(DEG_results_3$logFC) > 1) / nrow(DEG_results_3)) * 100,
    (sum(DEG_results_4$adj.P.Val < 0.05 & abs(DEG_results_4$logFC) > 1) / nrow(DEG_results_4)) * 100,
    (sum(DEG_results_5$adj.P.Val < 0.05 & abs(DEG_results_5$logFC) > 1) / nrow(DEG_results_5)) * 100
  ), 2)
)

print(deg_stats_padj_lfc)

```

#### Barplot
```{r}
# Preapare data for Bar Plot
deg_stats_bar_all <- data.frame(
  Group = rep(c("Before_vs_After", "R_Before_vs_After", "NR_Before_vs_After", "NR_vs_R_Before", "NR_vs_R_After"), each = 2),
  Regulation = rep(c("Upregulated", "Downregulated"), times = 5),
  Count = c(
    sum(DEG_results_1$logFC > 1 & DEG_results_1$adj.P.Val < 0.05),
    sum(DEG_results_1$logFC < -1 & DEG_results_1$adj.P.Val < 0.05),
    sum(DEG_results_2$logFC > 1 & DEG_results_2$adj.P.Val < 0.05),
    sum(DEG_results_2$logFC < -1 & DEG_results_2$adj.P.Val < 0.05),
    sum(DEG_results_3$logFC > 1 & DEG_results_3$adj.P.Val < 0.05),
    sum(DEG_results_3$logFC < -1 & DEG_results_3$adj.P.Val < 0.05),
    sum(DEG_results_4$logFC > 1 & DEG_results_4$adj.P.Val < 0.05),
    sum(DEG_results_4$logFC < -1 & DEG_results_4$adj.P.Val < 0.05),
    sum(DEG_results_5$logFC > 1 & DEG_results_5$adj.P.Val < 0.05),
    sum(DEG_results_5$logFC < -1 & DEG_results_5$adj.P.Val < 0.05)
  )
)
deg_stats_bar_all$Group <- factor(deg_stats_bar_all$Group, levels = c("Before_vs_After", "R_Before_vs_After", "NR_Before_vs_After", "NR_vs_R_Before", "NR_vs_R_After"))
# Build Bar Plot
barplot_degs_all <- ggplot(deg_stats_bar_all, aes(x = Group, y = Count, fill = Regulation)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.8), width = 0.7) +
  geom_text(aes(label = Count), vjust = -0.5, position = position_dodge(width = 0.8), size = 4) +
  scale_fill_manual(values = c("Upregulated" = "red", "Downregulated" = "blue")) +
  labs(
    x = "Groups",
    y = "Number of Significant Genes",
    fill = "Expression",
    title = "Bar Plot of Differentially Expressed Genes (Upregulated vs Downregulated)\nAll patients\nTotal number of genes: 23520"
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 14, face = "bold"),
    axis.title = element_text(size = 12),
    axis.text = element_text(size = 10),
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 10)
  )
```

```{r}
dir.create("../../../images/CD/train/DEGs/DEGs_All_Results", recursive = TRUE, showWarnings = FALSE) 

ggsave(filename = "../../../images/CD/train/DEGs/DEGs_All_Results/barplot_degs_all.png", plot = barplot_degs_all, width = 9, height = 5)
```


### Visualizations

#### Visualizations 1-3
```{r}
# Visualization of the first group

# Let's rewrite the table with DEGs
data_plot_deg_1 <- DEG_results_1
data_plot_deg_1$gene <- row.names(data_plot_deg_1)

# We determine which genes are differentially expressed
data_plot_deg_1$diffexpressed <- "Not significant"
data_plot_deg_1$diffexpressed[data_plot_deg_1$logFC > 1 & data_plot_deg_1$adj.P.Val < 0.05] <- "Upregulated"
data_plot_deg_1$diffexpressed[data_plot_deg_1$logFC < -1 & data_plot_deg_1$adj.P.Val < 0.05] <- "Downregulated"

# Select ONLY the TOP 10 Up and Top 10 Down genes
top_up <- data_plot_deg_1[data_plot_deg_1$diffexpressed == "Upregulated", ]
top_up <- top_up[order(-top_up$logFC), ][1:10, "gene"]

top_down <- data_plot_deg_1[data_plot_deg_1$diffexpressed == "Downregulated", ]
top_down <- top_down[order(top_down$logFC), ][1:10, "gene"]

# We only sign these genes
data_plot_deg_1$delabel <- ifelse(data_plot_deg_1$gene %in% c(top_up, top_down), data_plot_deg_1$gene, "")

#  Build Volcano Plot
plot_deg_1 <- ggplot(data = data_plot_deg_1, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  # Отсечки logFC
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  # Отсечка p-value
  geom_point(size = 2) +  # Размер точек
  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, max(-log10(DEG_results_1$adj.P.Val), na.rm = TRUE)), 
                  xlim = c(-max(abs(DEG_results_1$logFC), na.rm = TRUE), max(abs(DEG_results_1$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: Differentially Expressed Genes Before vs. After Treatment') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) + # Подписываем только топ-10 генов
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_deg_1)
```

```{r}
# Visualization 2
# Let's rewrite the table with DEGs
data_plot_deg_2 <- DEG_results_2
data_plot_deg_2$gene <- row.names(data_plot_deg_2)

# We determine which genes are differentially expressed
data_plot_deg_2$diffexpressed <- "Not significant"
data_plot_deg_2$diffexpressed[data_plot_deg_2$logFC > 1 & data_plot_deg_2$adj.P.Val < 0.05] <- "Upregulated"
data_plot_deg_2$diffexpressed[data_plot_deg_2$logFC < -1 & data_plot_deg_2$adj.P.Val < 0.05] <- "Downregulated"

# Select ONLY the TOP 10 Up and Top 10 Down genes
top_up <- data_plot_deg_2[data_plot_deg_2$diffexpressed == "Upregulated", ]
top_up <- top_up[order(-top_up$logFC), ][1:10, "gene"]

top_down <- data_plot_deg_2[data_plot_deg_2$diffexpressed == "Downregulated", ]
top_down <- top_down[order(top_down$logFC), ][1:10, "gene"]

# We only sign these genes
data_plot_deg_2$delabel <- ifelse(data_plot_deg_2$gene %in% c(top_up, top_down), data_plot_deg_2$gene, "")

#  Build a Volcano Plot
plot_deg_2 <- ggplot(data = data_plot_deg_2, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  # Отсечки logFC
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  # Отсечка p-value
  geom_point(size = 2) +  # Размер точек
  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, max(-log10(DEG_results_1$adj.P.Val)*2, na.rm = TRUE)), 
                  xlim = c(-max(abs(DEG_results_1$logFC), na.rm = TRUE), max(abs(DEG_results_1$logFC), na.rm = TRUE))*1.5) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: Differentially Expressed Genes Before vs. After Treatment (Responders)') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) + # Подписываем только топ-10 генов
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_deg_2)

```




```{r}
# Visualization 3
# Let's rewrite the table with DEGs
data_plot_deg_3 <- DEG_results_3
data_plot_deg_3$gene <- row.names(data_plot_deg_3)

# We determine which genes are differentially expressed
data_plot_deg_3$diffexpressed <- "Not significant"  # ВСЕ гены по умолчанию незначимые
data_plot_deg_3$diffexpressed[data_plot_deg_3$logFC > 1 & data_plot_deg_3$adj.P.Val < 0.05] <- "Upregulated"
data_plot_deg_3$diffexpressed[data_plot_deg_3$logFC < -1 & data_plot_deg_3$adj.P.Val < 0.05] <- "Downregulated"

# Let's check if there are any significant genes
if (sum(data_plot_deg_3$diffexpressed == "Upregulated") > 0) {
  top_up <- data_plot_deg_3[data_plot_deg_3$diffexpressed == "Upregulated", ]
  top_up_genes <- top_up[order(-top_up$logFC), ][1:min(10, nrow(top_up)), "gene"]
} else {
  top_up_genes <- character(0)  # Пустой вектор, если значимых генов нет
}

if (sum(data_plot_deg_3$diffexpressed == "Downregulated") > 0) {
  top_down <- data_plot_deg_3[data_plot_deg_3$diffexpressed == "Downregulated", ]
  top_down_genes <- top_down[order(top_down$logFC), ][1:min(10, nrow(top_down)), "gene"]
} else {
  top_down_genes <- character(0)  # Пустой вектор, если значимых генов нет
}

# We only sign the top 10 Up and Top 10 Down genes (if any)
data_plot_deg_3$delabel <- ifelse(data_plot_deg_3$gene %in% c(top_up_genes, top_down_genes), data_plot_deg_3$gene, "")

#  Строим Volcano Plot
plot_deg_3 <- ggplot(data = data_plot_deg_3, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  # Отсечки logFC
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  # Отсечка adj.P.Val
  geom_point(size = 2) +  # Размер точек
  scale_color_manual(values = c("Downregulated" = "#00AFBB", "Not significant" = "grey", "Upregulated" = "#bb0c00"),
                     labels = c("Not significant")) +
  coord_cartesian(ylim = c(0, max(-log10(na.omit(data_plot_deg_3$adj.P.Val))*2, na.rm = TRUE)), 
                  xlim = c(-max(abs(data_plot_deg_3$logFC), na.rm = TRUE), max(abs(data_plot_deg_3$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: Differentially Expressed Genes Before vs. After Treatment (Non-Responders)') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) + # Подписываем только топ-10 генов
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_deg_3)


```



#### Visualization 4-5
```{r}
# Visualization 4-5
# Let's rewrite the table with DEGs
data_plot_deg_4 <- DEG_results_4
data_plot_deg_4$gene <- rownames(data_plot_deg_4)

# We determine which genes are differentially expressed
data_plot_deg_4$diffexpressed <- "Not significant"
data_plot_deg_4$diffexpressed[data_plot_deg_4$logFC > 1 & data_plot_deg_4$adj.P.Val < 0.05] <- "Upregulated"
data_plot_deg_4$diffexpressed[data_plot_deg_4$logFC < -1 & data_plot_deg_4$adj.P.Val < 0.05] <- "Downregulated"

# We only sign the top 10 Up and Top 10 Down genes 
top_up <- data_plot_deg_4[data_plot_deg_4$diffexpressed == "Upregulated", ]
top_up <- top_up[order(-top_up$logFC), ][1:10, "gene"]

top_down <- data_plot_deg_4[data_plot_deg_4$diffexpressed == "Downregulated", ]
top_down <- top_down[order(top_down$logFC), ][1:10, "gene"]

data_plot_deg_4$delabel <- ifelse(data_plot_deg_4$gene %in% c(top_up, top_down), data_plot_deg_4$gene, "")

#  Строим Volcano Plot
plot_deg_4 <- ggplot(data = data_plot_deg_4, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  # Отсечки logFC
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  # Отсечка p-value
  geom_point(size = 2) +  # Размер точек
  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, max(-log10(DEG_results_4$adj.P.Val), na.rm = TRUE)), 
                  xlim = c(-max(abs(DEG_results_4$logFC), na.rm = TRUE), max(abs(DEG_results_4$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: Differentially Expressed Genes Before Treatment (Non-Responders vs. Responders)') +  # Закрываем скобку тут
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) +  # Подписываем только топ-10 генов
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_deg_4)


```


```{r}
# Visualization 5
# Let's rewrite the table with DEGs
data_plot_deg_5 <- DEG_results_5
data_plot_deg_5$gene <- rownames(data_plot_deg_5)

# We determine which genes are differentially expressed
data_plot_deg_5$diffexpressed <- "Not significant"
data_plot_deg_5$diffexpressed[data_plot_deg_5$logFC > 1 & data_plot_deg_5$adj.P.Val < 0.05] <- "Upregulated"
data_plot_deg_5$diffexpressed[data_plot_deg_5$logFC < -1 & data_plot_deg_5$adj.P.Val < 0.05] <- "Downregulated"

# Select ONLY the TOP 10 Up and Top 10 Down genes
top_up <- data_plot_deg_5[data_plot_deg_5$diffexpressed == "Upregulated", ]
top_up <- top_up[order(-top_up$logFC), ][1:10, "gene"]

top_down <- data_plot_deg_5[data_plot_deg_5$diffexpressed == "Downregulated", ]
top_down <- top_down[order(top_down$logFC), ][1:10, "gene"]

# We only sign these genes
data_plot_deg_5$delabel <- ifelse(data_plot_deg_5$gene %in% c(top_up, top_down), data_plot_deg_5$gene, "")

#  Build a Volcano Plot
plot_deg_5 <- ggplot(data = data_plot_deg_5, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  # Отсечки logFC
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  # Отсечка p-value
  geom_point(size = 2) +  # Размер точек
  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, max(-log10(DEG_results_5$adj.P.Val), na.rm = TRUE)), 
                  xlim = c(-max(abs(DEG_results_5$logFC), na.rm = TRUE), max(abs(DEG_results_5$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: Differentially Expressed Genes After Treatment (Non-Responders vs. Responders)') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) + # Подписываем только топ-10 генов
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_deg_5)

```
#### Saving the Volcano plots
```{r}
dir.create("../../../images/CD/train/DEGs/DEGs_All_Results", recursive = TRUE, showWarnings = FALSE) 

# Список графиков DEG
deg_list <- list(plot_deg_1, plot_deg_2, plot_deg_3, plot_deg_4, plot_deg_5)

# Filnames with prefix
file_names <- c(
  "1_Volcano_All_Before_After.png",  # Все пациенты ДО и ПОСЛЕ
  "2_Volcano_R_Before_After.png",    # Респондеры ДО и ПОСЛЕ
  "3_Volcano_NR_Before_After.png",   # Нереспондеры ДО и ПОСЛЕ
  "4_Volcano_R_NR_Before.png",       # Респондеры и Нереспондеры ДО
  "5_Volcano_R_NR_After.png"         # Респондеры и Нереспондеры ПОСЛЕ
)

for (i in seq_along(deg_list)) {
  file_path <- paste0("../../../images/CD/train/DEGs/DEGs_All_Results/", file_names[i])
  svg(file_path, width = 16, height = 10)
  print(deg_list[[i]])
  dev.off()
}

```


## Crohn's disease (CD):
Within this group there is also a division into Colon and Ileum. Since there is not a lot of data, we will analyze the data on Colon and Izoom together. 

Let's filter out the metadata and data with expression only with Crohn's disease:

```{r}
metadata_CD <- metadata[metadata$Disease == "CD", ]
mean_norm.expr_CD <- mean_norm.expr[, metadata_CD$Geo_accession]

# Convert the levels of the "Timeline" factor
metadata_CD$Timeline <- factor(metadata_CD$Timeline, 
                                          levels = c("Before first infliximab treatment", 
                                                     "After first infliximab treatment"),
                                          labels = c("before", "after"))

# Transform the levels of the "Response" factor
metadata_CD$Response <- factor(metadata_CD$Response, 
                                          levels = c("Yes", 
                                                     "No"))
```

We will store the expression data separately according to the metadata for each of the groups.:

1. The first comparison group: EVERYTHING before and after treatment. CD_all_beforeT_and_afterT
2. The second comparison group: ALL respondents before and after treatment. CD_R_beforeT_and_afterT
3. The third comparison group: ALL nonresponders before and after treatment. CD_NR_beforeT_and_afterT
4. The fourth comparison group: nonresponders BEFORE versus respondents BEFORE. CD_RNR_beforeT
5. The fifth comparison group: nonresponders AFTER versus respondents AFTER. CD_RNR_afterT

```{r}
m_CD_all_beforeT_and_afterT <- metadata_CD

m_CD_R_beforeT_and_afterT <- metadata_CD %>%
  dplyr::filter(Response == "Yes")

m_CD_NR_beforeT_and_afterT <- metadata_CD %>%
  dplyr::filter(Response == "No")

m_CD_RNR_beforeT <- metadata_CD %>%
  dplyr::filter(Timeline == "before")

m_CD_RNR_afterT <- metadata_CD %>%
  dplyr::filter(Timeline == "after")

```
Combine the expression data and metadata
```{r}
CD_1 <- mean_norm.expr[, m_CD_all_beforeT_and_afterT$Geo_accession]
CD_2 <- mean_norm.expr[, m_CD_R_beforeT_and_afterT$Geo_accession]
CD_3 <- mean_norm.expr[, m_CD_NR_beforeT_and_afterT$Geo_accession]
CD_4 <- mean_norm.expr[, m_CD_RNR_beforeT$Geo_accession]
CD_5 <- mean_norm.expr[, m_CD_RNR_afterT$Geo_accession]
```

### Design matrix 1-3:
```{r}
#Groups 1-3. Before and after

m_CD_all_beforeT_and_afterT$Timeline <- factor(m_CD_all_beforeT_and_afterT$Timeline, levels = c("before", "after"))
CD_design_1 <- model.matrix(~0 + m_CD_all_beforeT_and_afterT$Timeline)
colnames(CD_design_1) <- levels(m_CD_all_beforeT_and_afterT$Timeline)

m_CD_R_beforeT_and_afterT$Timeline <- factor(m_CD_R_beforeT_and_afterT$Timeline, levels = c("before", "after"))
CD_design_2 <- model.matrix(~0 + m_CD_R_beforeT_and_afterT$Timeline)
colnames(CD_design_2) <- levels(m_CD_R_beforeT_and_afterT$Timeline)

m_CD_NR_beforeT_and_afterT$Timeline <- factor(m_CD_NR_beforeT_and_afterT$Timeline, levels = c("before", "after"))
CD_design_3 <- model.matrix(~0 + m_CD_NR_beforeT_and_afterT$Timeline)
colnames(CD_design_3) <- levels(m_CD_NR_beforeT_and_afterT$Timeline)

# As far as I understand, ~0 allows us to create models without an intersept, and each group becomes a separate variable. 
```

### Design matrix 4-5:
Then we will build matrices for groups 4-5 to search for DEGs in the responder-nonresponder groups.

```{r}
# Groups 4-5. Non-responders-responders
m_CD_RNR_beforeT$Response <- factor(m_CD_RNR_beforeT$Response, levels = c("No", "Yes"))

CD_design_4 <- model.matrix(~0 + m_CD_RNR_beforeT$Response)
colnames(CD_design_4) <- levels(m_CD_RNR_beforeT$Response)


m_CD_RNR_afterT$Response <- factor(m_CD_RNR_afterT$Response, levels = c("No", "Yes"))

CD_design_5 <- model.matrix(~0 + m_CD_RNR_afterT$Response)
colnames(CD_design_5) <- levels(m_CD_RNR_afterT$Response)

```

### Contrast matrix 1-3:
Now we create "contrast matrices" -- makeContrasts. As far as I understand, we have defined groups in the design matrix, and makeContrast allows us to compare these groups.:

```{r}
# Groups 1-3

 Group 1: All samples before and after treatment
CD_contrast_1 <- makeContrasts(before - after, levels = CD_design_1)

# Группа 2: Только респондеры до и после
CD_contrast_2 <- makeContrasts(before - after, levels = CD_design_2)

# Группа 3: Только нереспондеры до и после
CD_contrast_3 <- makeContrasts(before - after, levels = CD_design_3)

```

### Contrast matrix 4-5:

```{r}
# Groups 4-5

# Group 4: Non-transponders vs respondents BEFORE treatment
CD_contrast_4 <- makeContrasts(No - Yes, levels = CD_design_4)

# Group 5: Non-responders vs POST-treatment responders
CD_contrast_5 <- makeContrasts(No - Yes, levels = CD_design_5)

```

We build models for each of the comparison groups:
```{r}
CD_fit_1 <- lmFit(CD_1, CD_design_1)
CD_fit_2 <- lmFit(CD_2, CD_design_2)
CD_fit_3 <- lmFit(CD_3, CD_design_3)

CD_fit_4 <- lmFit(CD_4, CD_design_4)
CD_fit_5 <- lmFit(CD_5, CD_design_5)
```
These linear models were built on the basis of a design matrix, but without comparing groups directly, so we use contrasts.fit() to "supplement" the resulting models with information about comparing groups (contrasting matrices):

```{r}
CD_fit2_1 <- contrasts.fit(CD_fit_1, CD_contrast_1)
CD_fit2_2 <- contrasts.fit(CD_fit_2, CD_contrast_2)
CD_fit2_3 <- contrasts.fit(CD_fit_3, CD_contrast_3)

CD_fit2_4 <- contrasts.fit(CD_fit_4, CD_contrast_4)
CD_fit2_5 <- contrasts.fit(CD_fit_5, CD_contrast_5)
```

We adjust the statistics and get the p-value values and others using eBayes

```{r}
CD_fit2_1 <- eBayes(CD_fit2_1)
CD_fit2_2 <- eBayes(CD_fit2_2)
CD_fit2_3 <- eBayes(CD_fit2_3)

CD_fit2_4 <- eBayes(CD_fit2_4)
CD_fit2_5 <- eBayes(CD_fit2_5)
```

### Results for all genes
We have quite a lot of data, so we will use the Benjamin-Hochberg correction to identify DEGs. In addition, we will save the files with the results separately.
```{r}
dir.create("../../../tables/CD/train/DEGs_CD_Results", recursive = TRUE, showWarnings = FALSE) 

CD_DEG_results_1 <- topTable(CD_fit2_1, coef = 1, adjust = "BH", number = Inf)
write.table(CD_DEG_results_1, "../../../tables/CD/train/DEGs_CD_Results/CD_DEG_results_1.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_DEG_results_2 <- topTable(CD_fit2_2, coef = 1, adjust = "BH", number = Inf)
write.table(CD_DEG_results_2, "../../../tables/CD/train/DEGs_CD_Results/CD_DEG_results_2.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_DEG_results_3 <- topTable(CD_fit2_3, coef = 1, adjust = "BH", number = Inf)
write.table(CD_DEG_results_3, "../../../tables/CD/train/DEGs_CD_Results/CD_DEG_results_3.txt", sep = "\t", quote = FALSE, row.names = TRUE)


CD_DEG_results_4 <- topTable(CD_fit2_4, coef = 1, adjust = "BH", number = Inf)
write.table(CD_DEG_results_4, "../../../tables/CD/train/DEGs_CD_Results/CD_DEG_results_4.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_DEG_results_5 <- topTable(CD_fit2_5, coef = 1, adjust = "BH", number = Inf)
write.table(CD_DEG_results_5, "../../../tables/CD/train/DEGs_CD_Results/CD_DEG_results_5.txt", sep = "\t", quote = FALSE, row.names = TRUE)


```
### DEGs

Let's filter out the significant genes and save the results.:

```{r}
CD_deg_sign_1 <- subset(CD_DEG_results_1, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(CD_deg_sign_1, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_1.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_deg_sign_2 <- subset(CD_DEG_results_2, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(CD_deg_sign_2, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_2.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_deg_sign_3 <- subset(CD_DEG_results_3, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(CD_deg_sign_3, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_3.txt", sep = "\t", quote = FALSE, row.names = TRUE)



CD_deg_sign_4 <- subset(CD_DEG_results_4, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(CD_deg_sign_4, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_4.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_deg_sign_5 <- subset(CD_DEG_results_5, adj.P.Val < 0.05 & abs(logFC) > 1)
write.table(CD_deg_sign_5, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_5.txt", sep = "\t", quote = FALSE, row.names = TRUE)

```

```{r}
# We will save ONLY SIGNIFICANT DEGs.(только adj.P.Val < 0.05)
CD_deg_sign_1_pval <- subset(CD_DEG_results_1, adj.P.Val < 0.05)
write.table(CD_deg_sign_1, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_1_pval.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_deg_sign_2_pval <- subset(CD_DEG_results_2, adj.P.Val < 0.05)
write.table(CD_deg_sign_2, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_2_pval.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_deg_sign_3_pval <- subset(CD_DEG_results_3, adj.P.Val < 0.05)
write.table(CD_deg_sign_3, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_3_pval.txt", sep = "\t", quote = FALSE, row.names = TRUE)



CD_deg_sign_4_pval <- subset(CD_DEG_results_4, adj.P.Val < 0.05)
write.table(CD_deg_sign_4, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_4_pval.txt", sep = "\t", quote = FALSE, row.names = TRUE)

CD_deg_sign_5_pval <- subset(CD_DEG_results_5, adj.P.Val < 0.05)
write.table(CD_deg_sign_5, "../../../tables/CD/train/DEGs_CD_Results/CD_deg_sign_5_pval.txt", sep = "\t", quote = FALSE, row.names = TRUE)
```


### Statistics

Statistics on adj.P.Val. Let's calculate how many DEGs adj.P.Val < 0.05 have

```{r}
deg_stats_padj_lfc_CD <- data.frame(
  Группа = c("Группа 1", "Группа 2", "Группа 3", "Группа 4", "Группа 5"),
  Всего_DEG = c(nrow(CD_DEG_results_1), nrow(CD_DEG_results_2), nrow(CD_DEG_results_3), nrow(CD_DEG_results_4), nrow(CD_DEG_results_5)),
  Значимые_DEG = c(sum(CD_DEG_results_1$adj.P.Val < 0.05 & abs(CD_DEG_results_1$logFC) > 1),
                   sum(CD_DEG_results_2$adj.P.Val < 0.05 & abs(CD_DEG_results_2$logFC) > 1),
                   sum(CD_DEG_results_3$adj.P.Val < 0.05 & abs(CD_DEG_results_3$logFC) > 1),
                   sum(CD_DEG_results_4$adj.P.Val < 0.05 & abs(CD_DEG_results_4$logFC) > 1),
                   sum(CD_DEG_results_5$adj.P.Val < 0.05 & abs(CD_DEG_results_5$logFC) > 1)),
  Процент_значимых = round(c(
    (sum(CD_DEG_results_1$adj.P.Val < 0.05 & abs(CD_DEG_results_1$logFC) > 1) / nrow(CD_DEG_results_1)) * 100,
    (sum(CD_DEG_results_2$adj.P.Val < 0.05 & abs(CD_DEG_results_2$logFC) > 1) / nrow(CD_DEG_results_2)) * 100,
    (sum(CD_DEG_results_3$adj.P.Val < 0.05 & abs(CD_DEG_results_3$logFC) > 1) / nrow(CD_DEG_results_3)) * 100,
    (sum(CD_DEG_results_4$adj.P.Val < 0.05 & abs(CD_DEG_results_4$logFC) > 1) / nrow(CD_DEG_results_4)) * 100,
    (sum(CD_DEG_results_5$adj.P.Val < 0.05 & abs(CD_DEG_results_5$logFC) > 1) / nrow(CD_DEG_results_5)) * 100
  ), 2)
)

print(deg_stats_padj_lfc_CD)


```
#### Barplot
```{r}

# Preparing data for the Bar Plot
deg_stats_bar <- data.frame(
  Group = rep(c("Before_vs_After", "R_Before_vs_After", "NR_Before_vs_After", "NR_vs_R_Before", "NR_vs_R_After"), each = 2),
  Regulation = rep(c("Upregulated", "Downregulated"), times = 5),
  Count = c(
    sum(CD_DEG_results_1$logFC > 1 & CD_DEG_results_1$adj.P.Val < 0.05),
    sum(CD_DEG_results_1$logFC < -1 & CD_DEG_results_1$adj.P.Val < 0.05),
    sum(CD_DEG_results_2$logFC > 1 & CD_DEG_results_2$adj.P.Val < 0.05),
    sum(CD_DEG_results_2$logFC < -1 & CD_DEG_results_2$adj.P.Val < 0.05),
    sum(CD_DEG_results_3$logFC > 1 & CD_DEG_results_3$adj.P.Val < 0.05),
    sum(CD_DEG_results_3$logFC < -1 & CD_DEG_results_3$adj.P.Val < 0.05),
    sum(CD_DEG_results_4$logFC > 1 & CD_DEG_results_4$adj.P.Val < 0.05),
    sum(CD_DEG_results_4$logFC < -1 & CD_DEG_results_4$adj.P.Val < 0.05),
    sum(CD_DEG_results_5$logFC > 1 & CD_DEG_results_5$adj.P.Val < 0.05),
    sum(CD_DEG_results_5$logFC < -1 & CD_DEG_results_5$adj.P.Val < 0.05)
  )
)
deg_stats_bar$Group <- factor(deg_stats_bar$Group, levels = c("Before_vs_After", "R_Before_vs_After", "NR_Before_vs_After", "NR_vs_R_Before", "NR_vs_R_After"))
# Построение Bar Plot
barplot_degs_CD <- ggplot(deg_stats_bar, aes(x = Group, y = Count, fill = Regulation)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.8), width = 0.7) +
  geom_text(aes(label = Count), vjust = -0.5, position = position_dodge(width = 0.8), size = 4) +
  scale_fill_manual(values = c("Upregulated" = "red", "Downregulated" = "blue")) +
  labs(
    x = "Groups",
    y = "Number of Significant Genes",
    fill = "Expression",
    title = "Bar Plot of Differentially Expressed Genes (Upregulated vs Downregulated)\nCrohn's Disease\nTotal number of genes: 23520"
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 14, face = "bold"),
    axis.title = element_text(size = 12),
    axis.text = element_text(size = 10),
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 10)
  )


```

```{r}
dir.create("../../../images/CD/train/DEGs/DEGs_CD_Results", recursive = TRUE, showWarnings = FALSE)
ggsave(filename = "../../../images/CD/train/DEGs/DEGs_CD_Results/barplot_degs_CD.png", plot = barplot_degs_CD, width = 9, height = 5)
```


### Visualizations 1-3
Visualization:
``{r}
# Visualization of the first group

# Let's rewrite the table with DEGs
data_plot_CD_DEG_results_1 <- CD_DEG_results_1
data_plot_CD_DEG_results_1$gene <- row.names(data_plot_CD_DEG_results_1)

# We determine which genes are differentially expressed
data_plot_CD_DEG_results_1$diffexpressed <- "Not significant"
data_plot_CD_DEG_results_1$diffexpressed[data_plot_CD_DEG_results_1$logFC > 1 & data_plot_CD_DEG_results_1$adj.P.Val < 0.05] <- "Upregulated"
data_plot_CD_DEG_results_1$diffexpressed[data_plot_CD_DEG_results_1$logFC < -1 & data_plot_CD_DEG_results_1$adj.P.Val < 0.05] <- "Downregulated"

# Select ONLY the TOP 10 Up and Top 10 Down genes
top_up <- data_plot_CD_DEG_results_1[data_plot_CD_DEG_results_1$diffexpressed == "Upregulated", ]
top_up <- top_up[order(-top_up$logFC), ][1:10, "gene"]

top_down <- data_plot_CD_DEG_results_1[data_plot_CD_DEG_results_1$diffexpressed == "Downregulated", ]
top_down <- top_down[order(top_down$logFC), ][1:10, "gene"]

# We only sign these genes
data_plot_CD_DEG_results_1$delabel <- ifelse(data_plot_CD_DEG_results_1$gene %in% c(top_up, top_down), data_plot_CD_DEG_results_1$gene, "")

plot_CD_DEG_results_1 <- ggplot(data = data_plot_CD_DEG_results_1, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  
  geom_point(size = 2) +  
  scale_color_manual(values = c("Downregulated" = "#00AFBB", "Not significant" = "grey", "Upregulated" = "#bb0c00")) +
  coord_cartesian(ylim = c(0, max(-log10(CD_DEG_results_1$adj.P.Val), na.rm = TRUE)), 
                  xlim = c(-max(abs(CD_DEG_results_1$logFC), na.rm = TRUE), max(abs(CD_DEG_results_1$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: CD. Differentially Expressed Genes Before vs. After Treatment') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) +  
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_CD_DEG_results_1)


```


```{r}
# Visualization of the second group

# Let's rewrite the table with DEGs
data_plot_CD_DEG_results_2 <- CD_DEG_results_2
data_plot_CD_DEG_results_2$gene <- row.names(data_plot_CD_DEG_results_2)

# We determine which genes are differentially expressed
data_plot_CD_DEG_results_2$diffexpressed <- "Not significant"
data_plot_CD_DEG_results_2$diffexpressed[data_plot_CD_DEG_results_2$logFC > 1 & data_plot_CD_DEG_results_2$adj.P.Val < 0.05] <- "Upregulated"
data_plot_CD_DEG_results_2$diffexpressed[data_plot_CD_DEG_results_2$logFC < -1 & data_plot_CD_DEG_results_2$adj.P.Val < 0.05] <- "Downregulated"

# Select ONLY the TOP 10 Up and Top 10 Down genes
top_up <- data_plot_CD_DEG_results_2[data_plot_CD_DEG_results_2$diffexpressed == "Upregulated", ]
top_up <- top_up[order(-top_up$logFC), ][1:10, "gene"]

top_down <- data_plot_CD_DEG_results_2[data_plot_CD_DEG_results_2$diffexpressed == "Downregulated", ]
top_down <- top_down[order(top_down$logFC), ][1:10, "gene"]

# We only sign these genes
data_plot_CD_DEG_results_2$delabel <- ifelse(data_plot_CD_DEG_results_2$gene %in% c(top_up, top_down), data_plot_CD_DEG_results_2$gene, "")

plot_CD_DEG_results_2 <- ggplot(data = data_plot_CD_DEG_results_2, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  
  geom_point(size = 2) +  
  scale_color_manual(values = c("Downregulated" = "#00AFBB", "Not significant" = "grey", "Upregulated" = "#bb0c00")) +
  coord_cartesian(ylim = c(0, max(-log10(CD_DEG_results_2$adj.P.Val), na.rm = TRUE)), 
                  xlim = c(-max(abs(CD_DEG_results_2$logFC), na.rm = TRUE), max(abs(CD_DEG_results_2$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: CD. Differentially Expressed Genes Before vs. After Treatment') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) +  
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_CD_DEG_results_2)


```


```{r}
# Visualization 3
# Overwriting the table with DEGs
data_plot_CD_DEG_results_3 <- CD_DEG_results_3
data_plot_CD_DEG_results_3$gene <- row.names(data_plot_CD_DEG_results_3)

# We determine which genes are differentially expressed
data_plot_CD_DEG_results_3$diffexpressed <- "Not significant"  # ВСЕ гены по умолчанию незначимые
data_plot_CD_DEG_results_3$diffexpressed[data_plot_CD_DEG_results_3$logFC > 1 & data_plot_CD_DEG_results_3$adj.P.Val < 0.05] <- "Upregulated"
data_plot_CD_DEG_results_3$diffexpressed[data_plot_CD_DEG_results_3$logFC < -1 & data_plot_CD_DEG_results_3$adj.P.Val < 0.05] <- "Downregulated"

# Let's check if there are any significant genes
if (sum(data_plot_CD_DEG_results_3$diffexpressed == "Upregulated") > 0) {
  top_up <- data_plot_CD_DEG_results_3[data_plot_CD_DEG_results_3$diffexpressed == "Upregulated", ]
  top_up_genes <- top_up[order(-top_up$logFC), ][1:min(10, nrow(top_up)), "gene"]
} else {
  top_up_genes <- character(0)  # Пустой вектор, если значимых генов нет
}

if (sum(data_plot_CD_DEG_results_3$diffexpressed == "Downregulated") > 0) {
  top_down <- data_plot_CD_DEG_results_3[data_plot_CD_DEG_results_3$diffexpressed == "Downregulated", ]
  top_down_genes <- top_down[order(top_down$logFC), ][1:min(10, nrow(top_down)), "gene"]
} else {
  top_down_genes <- character(0)  # Пустой вектор, если значимых генов нет
}

# We only sign the top 10 Up and Top 10 Down genes (if any)
data_plot_CD_DEG_results_3$delabel <- ifelse(data_plot_CD_DEG_results_3$gene %in% c(top_up_genes, top_down_genes), data_plot_CD_DEG_results_3$gene, "")

#  Build Volcano Plot
plot_CD_DEG_results_3 <- ggplot(data = data_plot_CD_DEG_results_3, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  # Отсечки logFC
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  # Отсечка adj.P.Val
  geom_point(size = 2) +  # Размер точек
  scale_color_manual(values = c("Downregulated" = "#00AFBB", "Not significant" = "grey", "Upregulated" = "#bb0c00"),
                     labels = c("Not significant")) +
  coord_cartesian(ylim = c(0, max(-log10(na.omit(data_plot_CD_DEG_results_3$adj.P.Val))*80000, na.rm = TRUE)), 
                  xlim = c(-max(abs(data_plot_CD_DEG_results_3$logFC), na.rm = TRUE), max(abs(data_plot_CD_DEG_results_3$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: CD. Differentially Expressed Genes Before vs. After Treatment (Non-Responders)') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) + # Подписываем только топ-10 генов
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_CD_DEG_results_3)

```

### Visualizations 4-5

```{r}
# Visualization of the fourth group

# Let's rewrite the table with DEGs
data_plot_CD_DEG_results_4 <- CD_DEG_results_4
data_plot_CD_DEG_results_4$gene <- row.names(data_plot_CD_DEG_results_4)

# We determine which genes are differentially expressed
data_plot_CD_DEG_results_4$diffexpressed <- "Not significant"
data_plot_CD_DEG_results_4$diffexpressed[data_plot_CD_DEG_results_4$logFC > 1 & data_plot_CD_DEG_results_4$adj.P.Val < 0.05] <- "Upregulated"
data_plot_CD_DEG_results_4$diffexpressed[data_plot_CD_DEG_results_4$logFC < -1 & data_plot_CD_DEG_results_4$adj.P.Val < 0.05] <- "Downregulated"

# Select ONLY the TOP 10 Up and Top 10 Down genes
top_up <- data_plot_CD_DEG_results_4[data_plot_CD_DEG_results_4$diffexpressed == "Upregulated", ]
top_up <- top_up[order(-top_up$logFC), ][1:10, "gene"]

top_down <- data_plot_CD_DEG_results_4[data_plot_CD_DEG_results_4$diffexpressed == "Downregulated", ]
top_down <- top_down[order(top_down$logFC), ][1:10, "gene"]

# We only sign these genes
data_plot_CD_DEG_results_4$delabel <- ifelse(data_plot_CD_DEG_results_4$gene %in% c(top_up, top_down), data_plot_CD_DEG_results_4$gene, "")

#  Build a Volcano Plot
plot_CD_DEG_results_4 <- ggplot(data = data_plot_CD_DEG_results_4, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  # Отсечки logFC
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  # Отсечка p-value
  geom_point(size = 2) +  # Размер точек
  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, max(-log10(CD_DEG_results_4$adj.P.Val), na.rm = TRUE)), 
                  xlim = c(-max(abs(CD_DEG_results_4$logFC), na.rm = TRUE), max(abs(CD_DEG_results_4$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: CD. Differentially Expressed Genes Before Treatment (Non-Responder vs. Responders)') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) + # Подписываем только топ-10 генов
  theme_minimal()+
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_CD_DEG_results_4)

```


```{r}
# Visualization of the fifth group

# Let's rewrite the table with DEGs
data_plot_CD_DEG_results_5 <- CD_DEG_results_5
data_plot_CD_DEG_results_5$gene <- row.names(data_plot_CD_DEG_results_5)

# We determine which genes are differentially expressed
data_plot_CD_DEG_results_5$diffexpressed <- "Not significant"
data_plot_CD_DEG_results_5$diffexpressed[data_plot_CD_DEG_results_5$logFC > 1 & data_plot_CD_DEG_results_5$adj.P.Val < 0.05] <- "Upregulated"
data_plot_CD_DEG_results_5$diffexpressed[data_plot_CD_DEG_results_5$logFC < -1 & data_plot_CD_DEG_results_5$adj.P.Val < 0.05] <- "Downregulated"

# Select ONLY the TOP 10 Up and Top 10 Down genes
top_up <- data_plot_CD_DEG_results_5[data_plot_CD_DEG_results_5$diffexpressed == "Upregulated", ]
top_up <- top_up[order(-top_up$logFC), ][1:10, "gene"]

top_down <- data_plot_CD_DEG_results_5[data_plot_CD_DEG_results_5$diffexpressed == "Downregulated", ]
top_down <- top_down[order(top_down$logFC), ][1:10, "gene"]

# We only sign these genes
data_plot_CD_DEG_results_5$delabel <- ifelse(data_plot_CD_DEG_results_5$gene %in% c(top_up, top_down), data_plot_CD_DEG_results_5$gene, "")

#  Build Volcano Plot
plot_CD_DEG_results_5 <- ggplot(data = data_plot_CD_DEG_results_5, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-1, 1), col = "black", linetype = 'dashed') +  # Отсечки logFC
  geom_hline(yintercept = -log10(0.05), col = "black", linetype = 'dashed') +  # Отсечка p-value
  geom_point(size = 2) +  # Размер точек
  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, max(-log10(CD_DEG_results_5$adj.P.Val), na.rm = TRUE)), 
                  xlim = c(-max(abs(CD_DEG_results_5$logFC), na.rm = TRUE), max(abs(CD_DEG_results_5$logFC), na.rm = TRUE))) +
  labs(color = 'Expression',
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"adj.p-value")) +
  ggtitle('Volcano Plot: CD. Differentially Expressed Genes After Treatment (Non-Responders vs. Responders)') +
  geom_text_repel(size = 4, max.overlaps = Inf, na.rm = TRUE) + # Подписываем только топ-10 генов
  theme_minimal() +
  theme(
    plot.title = element_text(size = 14, face = "bold"),      # Заголовок графика
    axis.title = element_text(size = 16),                     # Подписи осей
    axis.text = element_text(size = 14),                      # Метки осей
    legend.title = element_text(size = 14),                    # Заголовок легенды
    legend.text = element_text(size = 12)                      # Текст легенды
  )

print(plot_CD_DEG_results_5)


```
#### Saving the Volcano plots

```{r}
# Create a folder if it does not exist
dir.create("../../../images/CD/train/DEGs/DEGs_CD_Results", recursive = TRUE, showWarnings = FALSE)

# Список графиков DEG
deg_list <- list(plot_CD_DEG_results_1, plot_CD_DEG_results_2, plot_CD_DEG_results_3, plot_CD_DEG_results_4, plot_CD_DEG_results_5)

# PNG file names
file_names <- c(
  "1_CD_Volcano_All_Before_vs_After.png",
  "2_CD_Volcano_R_Before_vs_After.png",
  "3_CD_Volcano_NR_Before_vs_After.png",
  "4_CD_Volcano_NR_vs_R_Before.png",
  "5_CD_Volcano_NR_vs_R_After.png"
)

# Saving graphs in PNG
for (i in seq_along(deg_list)) {
  file_path <- paste0("../../../images/CD/train/DEGs/DEGs_CD_Results/", file_names[i])
  ggsave(filename = file_path, plot = deg_list[[i]], device = "png", width = 9, height = 6, dpi = 300)
}
```


# GSEA

First, we will conduct GSEA groups for all diseases, and then for the Crown.

## GSEA All patients
1. GSEA all dataset:
Its recomended to convert Gene Symbol into EntrezID, but here about 25% of genes can not be analized according to warning, we will use keyType - SYMBOL.

Build ranging by -log10(P.Value) * sign(logFC) (logFC or t-statistic are also good options) gene lists:
```{r}
# DEG_results_1
gene_list_1 <- with(DEG_results_1, -log10(P.Value) * sign(logFC))
names(gene_list_1) <- rownames(DEG_results_1)
gene_list_1 <- sort(gene_list_1, decreasing = TRUE)

organism <- "org.Hs.eg.db"
library(organism, character.only = TRUE)

# DEG_results_2
gene_list_2 <- with(DEG_results_2, -log10(P.Value) * sign(logFC))
names(gene_list_2) <- rownames(DEG_results_2)
gene_list_2 <- sort(gene_list_2, decreasing = TRUE)

# DEG_results_3
gene_list_3 <- with(DEG_results_3, -log10(P.Value) * sign(logFC))
names(gene_list_3) <- rownames(DEG_results_3)
gene_list_3 <- sort(gene_list_3, decreasing = TRUE)

# DEG_results_4
gene_list_4 <- with(DEG_results_4, -log10(P.Value) * sign(logFC))
names(gene_list_4) <- rownames(DEG_results_4)
gene_list_4 <- sort(gene_list_4, decreasing = TRUE)

# DEG_results_5
gene_list_5 <- with(DEG_results_5, -log10(P.Value) * sign(logFC))
names(gene_list_5) <- rownames(DEG_results_5)
gene_list_5 <- sort(gene_list_5, decreasing = TRUE)

```
GSEA (background list is not needed):
```{r}
# GSEA gene_list_1
gsea_results_1 <- gseGO(
  geneList      = gene_list_1,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

# GSEA gene_list_2
gsea_results_2 <- gseGO(
  geneList      = gene_list_2,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

# GSEA gene_list_3
gsea_results_3 <- gseGO(
  geneList      = gene_list_3,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

# GSEA gene_list_4
gsea_results_4 <- gseGO(
  geneList      = gene_list_4,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

# GSEA gene_list_5
gsea_results_5 <- gseGO(
  geneList      = gene_list_5,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)


```
### Visualization
Visualization (difference in p.adj is small)
```{r}
# https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html

# Dotplot gsea_results_1
dotplot_1 <- dotplot(gsea_results_1, showCategory = 5, split = ".sign") +
   facet_grid(. ~ .sign) +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
)

# Dotplot gsea_results_2
dotplot_2 <- dotplot(gsea_results_2, showCategory = 5, split = ".sign") +
   facet_grid(. ~ .sign) +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
   )

# Dotplot gsea_results_3
dotplot_3 <- dotplot(gsea_results_3, showCategory = 5, split = ".sign") +
   facet_grid(. ~ .sign) +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
   )

# Dotplot gsea_results_4
dotplot_4 <- dotplot(gsea_results_4, showCategory = 5, split = ".sign") +
   facet_grid(. ~ .sign) +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
   )

# Dotplot gsea_results_5
dotplot_5 <- dotplot(gsea_results_5, showCategory = 5, split = ".sign") +
   facet_grid(. ~ .sign) +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
   )


```


```{r}
dir.create("../../../images/CD/train/GSEA/GSEA_All_Results", recursive = TRUE, showWarnings = FALSE)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/dotplot_1.png", plot = dotplot_1, 
       width = 10, height = 8,  dpi = 300)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/dotplot_2.png", plot = dotplot_2, 
       width = 10, height = 8,  dpi = 300)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/dotplot_3.png", plot = dotplot_3, 
       width = 10, height = 8,  dpi = 300)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/dotplot_4.png", plot = dotplot_4, 
       width = 10, height = 8,  dpi = 300)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/dotplot_5.png", plot = dotplot_5, 
       width = 10, height = 8, dpi = 300)

```



```{r}
# Cnetplot gsea_results_1
cnetplot_1 <- cnetplot(gsea_results_1, showCategory = 3)

# Cnetplot gsea_results_2
cnetplot_2 <- cnetplot(gsea_results_2, showCategory = 3)

# Cnetplot gsea_results_3
cnetplot_3 <- cnetplot(gsea_results_3, showCategory = 3)

# Cnetplot gsea_results_4
cnetplot_4 <- cnetplot(gsea_results_4, showCategory = 3)

# Cnetplot gsea_results_5
cnetplot_5 <- cnetplot(gsea_results_5, showCategory = 3)

```

```{r}
# Saving cnetplot_1 - cnetplot_5 in PNG
ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/cnetplot_1.jpg", plot = cnetplot_1, 
       width = 16, height = 10, dpi = 300)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/cnetplot_2.jpg", plot = cnetplot_2, 
       width = 16, height = 10, dpi = 300)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/cnetplot_3.jpg", plot = cnetplot_3, 
       width = 16, height = 10, dpi = 300)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/cnetplot_4.jpg", plot = cnetplot_4, 
       width = 16, height = 10, dpi = 300)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_All_Results/cnetplot_5.jpg", plot = cnetplot_5, 
       width = 16, height = 10, dpi = 300)

```


## ORA All patients
ORA:

```{r}
# Take significant DEGs с padj < 0.05 и |logFC| > 1
deg_list_1 <- rownames(DEG_results_1[DEG_results_1$adj.P.Val < 0.05 & abs(DEG_results_1$logFC) > 1, ])
deg_list_2 <- rownames(DEG_results_2[DEG_results_2$adj.P.Val < 0.05 & abs(DEG_results_2$logFC) > 1, ])
# deg_list_3 <- rownames(DEG_results_3[DEG_results_3$adj.P.Val < 0.05 & abs(DEG_results_3$logFC) > 1, ])
deg_list_4 <- rownames(DEG_results_4[DEG_results_4$adj.P.Val < 0.05 & abs(DEG_results_4$logFC) > 1, ])
deg_list_5 <- rownames(DEG_results_5[DEG_results_5$adj.P.Val < 0.05 & abs(DEG_results_5$logFC) > 1, ])

```

Define background gene list. Gene lust on the array. Can be taken from DEG_results_
```{r}
universe <- rownames(DEG_results_1)  # Все гены, присутствующие в анализе

```

```{r}
# ORAfor deg_list_1
ora_results_1 <- enrichGO(
  gene          = deg_list_1,
  universe      = universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

# ORA for deg_list_2
ora_results_2 <- enrichGO(
  gene          = deg_list_2,
  universe      = universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

# ORA for deg_list_3
# ora_results_3 <- enrichGO(
#   gene          = deg_list_3,
#   universe      = universe,
#   OrgDb         = org.Hs.eg.db,
#   keyType       = "SYMBOL",
#   ont           = "BP",
#   pAdjustMethod = "BH",
#   pvalueCutoff  = 0.05
# )

# ORA for deg_list_4
ora_results_4 <- enrichGO(
  gene          = deg_list_4,
  universe      = universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

# ORA for deg_list_5
ora_results_5 <- enrichGO(
  gene          = deg_list_5,
  universe      = universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

```
### Visualization
```{r}

# Dotplot для ora_results_1
ora_dotplot_1 <- dotplot(ora_results_1, showCategory=10)

# Dotplot для ora_results_2
ora_dotplot_2 <- dotplot(ora_results_2, showCategory=10)

# Dotplot для ora_results_3
#ora_dotplot_3 <- dotplot(ora_results_3, showCategory=10)

# Dotplot для ora_results_4
ora_dotplot_4 <- dotplot(ora_results_4, showCategory=10)

# Dotplot для ora_results_5
ora_dotplot_5 <- dotplot(ora_results_5, showCategory=10)
```

```{r}
dir.create("../../../images/CD/train/ORA/ORA_All_Results", recursive = TRUE, showWarnings = FALSE)
# Saving dotplot_1 - dotplot_5 in PNG
ggsave(filename = "../../../images/CD/train/ORA/ORA_All_Results/ora_dotplot_1.png", plot = ora_dotplot_1, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_All_Results/ora_dotplot_2.png", plot = ora_dotplot_2, width = 16, height = 10, dpi = 300)
#ggsave(filename = "ora_dotplot_3.png", plot = ora_dotplot_3, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_All_Results/ora_dotplot_4.png", plot = ora_dotplot_4, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_All_Results/ora_dotplot_5.png", plot = ora_dotplot_5, width = 16, height = 10, dpi = 300)



# Cnetplot for ora_results_1
ora_cnetplot_1 <- cnetplot(ora_results_1, showCategory = 3)

# Cnetplot for ora_results_2
ora_cnetplot_2 <- cnetplot(ora_results_2, showCategory = 3)

# Cnetplot for ora_results_3
#ora_cnetplot_3 <- cnetplot(ora_results_3, showCategory = 3)

# Cnetplot for ora_results_4
ora_cnetplot_4 <- cnetplot(ora_results_4, showCategory = 3)

# Cnetplot for ora_results_5
ora_cnetplot_5 <- cnetplot(ora_results_5, showCategory = 3)

# Saving cnetplot_1 - cnetplot_5 in JPG
ggsave(filename = "../../../images/CD/train/ORA/ORA_All_Results/ora_cnetplot_1.jpg", plot = ora_cnetplot_1, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_All_Results/ora_cnetplot_2.jpg", plot = ora_cnetplot_2, width = 16, height = 10, dpi = 300)
#ggsave(filename = "ora_cnetplot_3.jpg", plot = ora_cnetplot_3, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_All_Results/ora_cnetplot_4.jpg", plot = ora_cnetplot_4, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_All_Results/ora_cnetplot_5.jpg", plot = ora_cnetplot_5, width = 16, height = 10, dpi = 300)

```

## GSEA CD
Ranging by -log10(P.Value) * sign(logFC) genes for CD groups:
```{r}
CD_gene_list_1 <- with(CD_DEG_results_1, -log10(P.Value) * sign(logFC))
names(CD_gene_list_1) <- rownames(CD_DEG_results_1)
CD_gene_list_1 <- sort(CD_gene_list_1, decreasing = TRUE)

CD_gene_list_2 <- with(CD_DEG_results_2, -log10(P.Value) * sign(logFC))
names(CD_gene_list_2) <- rownames(CD_DEG_results_2)
CD_gene_list_2 <- sort(CD_gene_list_2, decreasing = TRUE)

CD_gene_list_3 <- with(CD_DEG_results_3, -log10(P.Value) * sign(logFC))
names(CD_gene_list_3) <- rownames(CD_DEG_results_3)
CD_gene_list_3 <- sort(CD_gene_list_3, decreasing = TRUE)

CD_gene_list_4 <- with(CD_DEG_results_4, -log10(P.Value) * sign(logFC))
names(CD_gene_list_4) <- rownames(CD_DEG_results_4)
CD_gene_list_4 <- sort(CD_gene_list_4, decreasing = TRUE)

CD_gene_list_5 <- with(CD_DEG_results_5, -log10(P.Value) * sign(logFC))
names(CD_gene_list_5) <- rownames(CD_DEG_results_5)
CD_gene_list_5 <- sort(CD_gene_list_5, decreasing = TRUE)
```

GSEA:
```{r}
CD_gsea_results_1 <- gseGO(
  geneList      = CD_gene_list_1,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

CD_gsea_results_2 <- gseGO(
  geneList      = CD_gene_list_2,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

CD_gsea_results_3 <- gseGO(
  geneList      = CD_gene_list_3,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

CD_gsea_results_4 <- gseGO(
  geneList      = CD_gene_list_4,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

CD_gsea_results_5 <- gseGO(
  geneList      = CD_gene_list_5,  
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL", 
  ont           = "ALL",
  nPerm         = 1000,
  minGSSize     = 3,
  maxGSSize     = 500,
  verbose       = TRUE,
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)
```
### Visualization:

```{r}
# Dotplot для CD_gsea_results_1
CD_gsea_dotplot_1 <- dotplot(CD_gsea_results_1, showCategory = 5, split = ".sign") +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
   ) +
   ggtitle("CD. Before vs. After")  # Добавляем заголовок

# Dotplot для CD_gsea_results_2
CD_gsea_dotplot_2 <- dotplot(CD_gsea_results_2, showCategory = 5, split = ".sign") +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
   ) +
   ggtitle("CD. Responders: Before vs. After")  # Добавляем заголовок

# Dotplot для CD_gsea_results_3
# #CD_gsea_dotplot_3 <- dotplot(CD_gsea_results_3, showCategory = 5, split = ".sign") +
#    theme(axis.text.x = element_text(angle = 45, hjust = 1),
#          axis.text.y = element_text(size = 10)
#    )

# Dotplot для CD_gsea_results_4
CD_gsea_dotplot_4 <- dotplot(CD_gsea_results_4, showCategory = 5, split = ".sign") +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
   ) +
   ggtitle("CD. After: Non-Responders vs. Responders")  # Добавляем заголовок

# Dotplot для CD_gsea_results_5
CD_gsea_dotplot_5 <- dotplot(CD_gsea_results_5, showCategory = 5, split = ".sign") +
   theme(axis.text.x = element_text(angle = 45, hjust = 1),
         axis.text.y = element_text(size = 10)
   ) +
   ggtitle("CD. Before: Non-Responders vs. Responders")  # Добавляем заголовок
```

```{r}
dir.create("../../../images/CD/train/GSEA/GSEA_CD_Results", recursive = TRUE, showWarnings = FALSE)

# Save dotplot_1 - dotplot_5 in PNG
ggsave(filename = "../../../images/CD/train/GSEA/GSEA_CD_Results/CD_dotplot_1.png", plot = CD_gsea_dotplot_1, 
       width = 10, height = 8,  dpi = 600)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_CD_Results/CD_dotplot_2.png", plot = CD_gsea_dotplot_2, 
       width = 10, height = 8,  dpi = 600)

# ggsave(filename = "CD_dotplot_3.png", plot = CD_gsea_dotplot_3, 
#        width = 10, height = 8,  dpi = 600)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_CD_Results/CD_dotplot_4.png", plot = CD_gsea_dotplot_4, 
       width = 10, height = 8,  dpi = 600)

ggsave(filename = "../../../images/CD/train/GSEA/GSEA_CD_Results/CD_dotplot_5.png", plot = CD_gsea_dotplot_5, 
       width = 10, height = 8, dpi = 600)
```



```{r}
# Cnetplot for ora_results_1
CD_gsea_cnetplot_1 <- cnetplot(CD_gsea_results_1, showCategory = 3)

# Cnetplot for ora_results_2
CD_gsea_cnetplot_2 <- cnetplot(CD_gsea_results_2, showCategory = 3)

# Cnetplot for ora_results_3
#ora_cnetplot_3 <- cnetplot(ora_results_3, showCategory = 3)

# Cnetplot for ora_results_4
CD_gsea_cnetplot_4 <- cnetplot(CD_gsea_results_4, showCategory = 3)

# Cnetplot for ora_results_5
CD_gsea_cnetplot_5 <- cnetplot(CD_gsea_results_5, showCategory = 3)


# Save cnetplot_1 - cnetplot_5 in JPG
ggsave(filename = "../../../images/CD/train/GSEA/GSEA_CD_Results/CD_gsea_cnetplot_1.jpg", plot = CD_gsea_cnetplot_1, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/GSEA/GSEA_CD_Results/CD_gsea_cnetplot_2.jpg", plot = CD_gsea_cnetplot_2, width = 16, height = 10, dpi = 300)
#ggsave(filename = "ora_cnetplot_3.jpg", plot = ora_cnetplot_3, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/GSEA/GSEA_CD_Results/CD_gsea_cnetplot_4.jpg", plot = CD_gsea_cnetplot_4, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/GSEA/GSEA_CD_Results/CD_gsea_cnetplot_5.jpg", plot = CD_gsea_cnetplot_5, width = 16, height = 10, dpi = 300)
```



## ORA CD:
```{r}
CD_deg_list_1 <- rownames(CD_DEG_results_1[CD_DEG_results_1$adj.P.Val < 0.05 & abs(CD_DEG_results_1$logFC) > 1, ])
CD_deg_list_2 <- rownames(CD_DEG_results_2[CD_DEG_results_2$adj.P.Val < 0.05 & abs(CD_DEG_results_2$logFC) > 1, ])
#CD_deg_list_3 <- rownames(CD_DEG_results_3[CD_DEG_results_3$adj.P.Val < 0.05 & abs(CD_DEG_results_3$logFC) > 1, ])
CD_deg_list_4 <- rownames(CD_DEG_results_4[CD_DEG_results_4$adj.P.Val < 0.05 & abs(CD_DEG_results_4$logFC) > 1, ])
CD_deg_list_5 <- rownames(CD_DEG_results_5[CD_DEG_results_5$adj.P.Val < 0.05 & abs(CD_DEG_results_5$logFC) > 1, ])

```

```{r}
# Define the background list universe (all changed genes)
CD_universe <- unique(c(rownames(CD_DEG_results_1), rownames(CD_DEG_results_2), 
                        rownames(CD_DEG_results_3), rownames(CD_DEG_results_4), 
                        rownames(CD_DEG_results_5)))
```

```{r}
CD_ora_results_1 <- enrichGO(
  gene          = CD_deg_list_1,
  universe      = CD_universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

CD_ora_results_2 <- enrichGO(
  gene          = CD_deg_list_2,
  universe      = CD_universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

# #CD_ora_results_3 <- enrichGO(
#   gene          = CD_deg_list_3,
#   universe      = CD_universe,
#   OrgDb         = org.Hs.eg.db,
#   keyType       = "SYMBOL",
#   ont           = "BP",
#   pAdjustMethod = "BH",
#   pvalueCutoff  = 0.05
# )

CD_ora_results_4 <- enrichGO(
  gene          = CD_deg_list_4,
  universe      = CD_universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

CD_ora_results_5 <- enrichGO(
  gene          = CD_deg_list_5,
  universe      = CD_universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)


```

### Visualization:
```{r}
# Dotplot for CD_ora_results_1
CD_ora_dotplot_1 <- dotplot(CD_ora_results_1, showCategory=10)

# Dotplot for CD_ora_results_2
CD_ora_dotplot_2 <- dotplot(CD_ora_results_2, showCategory=10)

# Dotplot for CD_ora_results_3
#CD_ora_dotplot_3 <- dotplot(CD_ora_results_3, showCategory=10)

# Dotplot for CD_ora_results_4
CD_ora_dotplot_4 <- dotplot(CD_ora_results_4, showCategory=10)

# Dotplot for CD_ora_results_5
CD_ora_dotplot_5 <- dotplot(CD_ora_results_5, showCategory=10)


dir.create("../../../images/CD/train/ORA/ORA_CD_Results", recursive = TRUE, showWarnings = FALSE)
# Saving dotplot_1 - dotplot_5 in PNG
ggsave(filename = "../../../images/CD/train/ORA/ORA_CD_Results/CD_ora_dotplot_1.png", plot = CD_ora_dotplot_1, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_CD_Results/CD_ora_dotplot_2.png", plot = CD_ora_dotplot_2, width = 16, height = 10, dpi = 300)
#ggsave(filename = "CD_ora_dotplot_3.png", plot = CD_ora_dotplot_3, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_CD_Results/CD_ora_dotplot_4.png", plot = CD_ora_dotplot_4, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_CD_Results/CD_ora_dotplot_5.png", plot = CD_ora_dotplot_5, width = 16, height = 10, dpi = 300)

# Cnetplot for CD_ora_results_1
CD_ora_cnetplot_1 <- cnetplot(CD_ora_results_1, showCategory = 3)

# Cnetplot for CD_ora_results_2
CD_ora_cnetplot_2 <- cnetplot(CD_ora_results_2, showCategory = 3)

# Cnetplot for CD_ora_results_3
#CD_ora_cnetplot_3 <- cnetplot(CD_ora_results_3, showCategory = 3)

# Cnetplot for CD_ora_results_4
CD_ora_cnetplot_4 <- cnetplot(CD_ora_results_4, showCategory = 3)

# Cnetplot для CD_ora_results_5
CD_ora_cnetplot_5 <- cnetplot(CD_ora_results_5, showCategory = 3)

# Save cnetplot_1 - cnetplot_5 into JPG
ggsave(filename = "../../../images/CD/train/ORA/ORA_CD_Results/CD_ora_cnetplot_1.jpg", plot = CD_ora_cnetplot_1, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_CD_Results/CD_ora_cnetplot_2.jpg", plot = CD_ora_cnetplot_2, width = 16, height = 10, dpi = 300)
#ggsave(filename = "CD_ora_cnetplot_3.jpg", plot = CD_ora_cnetplot_3, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_CD_Results/CD_ora_cnetplot_4.jpg", plot = CD_ora_cnetplot_4, width = 16, height = 10, dpi = 300)
ggsave(filename = "../../../images/CD/train/ORA/ORA_CD_Results/CD_ora_cnetplot_5.jpg", plot = CD_ora_cnetplot_5, width = 16, height = 10, dpi = 300)

```


Save results:
```{r}
dir.create("../../../tables/CD/train/GSEA_All_Results", recursive = TRUE, showWarnings = FALSE)
write.table(gsea_results_1, "../../../tables/CD/train/GSEA_All_Results/gsea_results_1.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(gsea_results_2, "../../../tables/CD/train/GSEA_All_Results/gsea_results_2.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(gsea_results_3, "../../../tables/CD/train/GSEA_All_Results/gsea_results_3.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(gsea_results_4, "../../../tables/CD/train/GSEA_All_Results/gsea_results_4.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(gsea_results_5, "../../../tables/CD/train/GSEA_All_Results/gsea_results_5.txt", sep = "\t", quote = FALSE, row.names = FALSE)

dir.create("../../../tables/CD/train/GSEA_CD_Results", recursive = TRUE, showWarnings = FALSE)
write.table(CD_gsea_results_1, "../../../tables/CD/train/GSEA_CD_Results/CD_gsea_results_1.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(CD_gsea_results_2, "../../../tables/CD/train/GSEA_CD_Results/CD_gsea_results_2.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(CD_gsea_results_3, "../../../tables/CD/train/GSEA_CD_Results/CD_gsea_results_3.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(CD_gsea_results_4, "../../../tables/CD/train/GSEA_CD_Results/CD_gsea_results_4.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(CD_gsea_results_5, "../../../tables/CD/train/GSEA_CD_Results/CD_gsea_results_5.txt", sep = "\t", quote = FALSE, row.names = FALSE)

```

# WGCNA


## Top 7500 genes WGCNA
Take 7500 variable gees, to avoid noise effects.

mean_norm.expr_CD keep the data about CD patients. Rewrite it into table to not looose it mean_norm.expr_CD
```{r}
CD_WGCNA_data <- mean_norm.expr_CD # данные об экспрессии
CD_WGCNA_metadata <- metadata_CD # метаданные
```

Transpose the matrix
```{r}
input_mat <- t(CD_WGCNA_data)

```

Choose top 7500 genes
```{r}
# colnames(input_mat) <- input_mat[1, ]
# input_mat <- input_mat[-1, ]
input_mat_filtered <- input_mat[rownames(input_mat) %in% m_CD_RNR_beforeT$Geo_accession, ]

gene_variances <- apply(input_mat_filtered, 2, var, na.rm = TRUE)
top7500_genes <- names(sort(gene_variances, decreasing = TRUE)[1:7500])
input_mat_top7500 <- input_mat_filtered[, top7500_genes, drop = FALSE]
dim(input_mat_top7500)
```

```{r}
allowWGCNAThreads()          # allow multi-threading (optional)

# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to = 20, by = 2))

# Call the network topology analysis function
sft_7500 = pickSoftThreshold(
  input_mat_top7500,             # <= Input data
  powerVector = powers,
  verbose = 5
)

par(mfrow = c(1,2));
cex1 = 0.9;

plot(sft_7500$fitIndices[, 1],
     -sign(sft_5000$fitIndices[, 3]) * sft_7500$fitIndices[, 2],
     xlab = "Soft Threshold (power)",
     ylab = "Scale Free Topology Model Fit, signed R^2",
     main = paste("Scale independence")
)
text(sft_7500$fitIndices[, 1],
     -sign(sft_7500$fitIndices[, 3]) * sft_7500$fitIndices[, 2],
     labels = powers, cex = cex1, col = "red"
)
abline(h = 0.85, col = "red")
plot(sft_7500$fitIndices[, 1],
     sft_7500$fitIndices[, 5],
     xlab = "Soft Threshold (power)",
     ylab = "Mean Connectivity",
     type = "n",
     main = paste("Mean connectivity")
)
text(sft_7500$fitIndices[, 1],
     sft_7500$fitIndices[, 5],
     labels = powers,
     cex = cex1, col = "red")

```


```{r}
picked_power_7500 = 7
temp_cor <- cor       
cor <- WGCNA::cor         # Force it to use WGCNA cor function (fix a namespace conflict issue)
netwk_7500 <- blockwiseModules(input_mat_top7500,                # <= input here
                               
                               # == Adjacency Function ==
                               power = picked_power_7500,                # <= power here
                               networkType = "signed",
                               
                               # == Tree and Block Options ==
                               deepSplit = 2,
                               pamRespectsDendro = F,
                               # detectCutHeight = 0.75,
                               minModuleSize = 30,
                               maxBlockSize = 4000,
                               
                               # == Module Adjustments ==
                               reassignThreshold = 0,
                               mergeCutHeight = 0.25,
                               
                               # == TOM == Archive the run results in TOM file (saves time)
                               saveTOMs = T,
                               saveTOMFileBase = "ER",
                               
                               # == Output Options
                               numericLabels = T,
                               verbose = 3)

cor <- temp_cor     # Return cor function to original namespace
```
```{r}
# Convert labels to colors for plotting
mergedColors_7500 = labels2colors(netwk_7500$colors)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(
  netwk_7500$dendrograms[[1]],
  mergedColors_7500[netwk_7500$blockGenes[[1]]],
  "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05 )
```

```{r}
# The table contains the gene and the corresponding module color.
module_df_7500 <- data.frame(
  gene_id = names(netwk_7500$colors),
  colors = labels2colors(netwk_7500$colors)
)

module_df_7500[1:5,]

```

```{r}
batch <- m_CD_RNR_beforeT$Response

batch <- as.data.frame(batch)

# One-Hot Encoding
batch_numeric <- model.matrix(~ . - 1, data = batch)

# Convert into data.frame (or it will be a matrix)
batch_numeric <- as.data.frame(batch_numeric)


module.trait.corr <- cor(netwk_7500$MEs, batch_numeric, use = 'p')
module.trait.corr.pvals <- corPvalueStudent(module.trait.corr, nrow(input_mat))

rownames(batch_numeric) <- m_CD_RNR_beforeT$Geo_accession

```
### Visualize
```{r}
heatmap.data <- merge(netwk_7500$MEs, batch_numeric, by = 'row.names')

head(heatmap.data)

library(tidyverse)
heatmap.data <- heatmap.data %>% 
  column_to_rownames(var = 'Row.names')

heatmap.data <- heatmap.data[, c(19, 20, 1:18)]



wgcna_plot_7500 <- CorLevelPlot(heatmap.data,
             x = names(heatmap.data)[2:1],
             y = names(heatmap.data)[3:20],
             col = c("blue1", "skyblue", "white", "pink", "red"))
wgcna_plot_7500 <- update(wgcna_plot_7500,
  main = "WGCNA Correlation Heatmap. CD. NR vs. R. Top 7500 genes.",
  xlab = "Traits",
  ylab = "Module Eigengenes"
)


dir.create("../../../images/CD/train/WGCNA_CD_7500_Results", recursive = TRUE, showWarnings = FALSE)

png("../../../images/CD/train/WGCNA_CD_7500_Results/WGCNA_plot.png", width = 10, height = 8, units = "in", res = 600)
print(wgcna_plot_7500)  # обязательно print!
dev.off()

```

Choose modules ME0, ME3 и ME8 for GO


```{r}
module_to_MEs <- data.frame(
  MEs_Name = colnames(netwk_7500$MEs),
  ModuleColor = unique(mergedColors_7500)
)

print(module_to_MEs)

```

### GO for modules
```{r}
dir.create("../../../tables/CD/train/WGCNA_Modules_Genes", recursive = TRUE, showWarnings = FALSE)
# ME0
ME0_df <- subset(module_df_7500, colors == "grey60")
ME0_genes <- ME0_df$gene_id

me0 <- as.data.frame(ME0_genes)
write_delim(me0, '../../../tables/CD/train/WGCNA_Modules_Genes/ME0_genes.txt', delim = '\t')

# ME2
ME2_df <- subset(module_df_7500, colors == "yellow")
ME2_genes <- ME2_df$gene_id

me2 <- as.data.frame(ME2_genes)
write_delim(me2, '../../../tables/CD/train/WGCNA_Modules_Genes/ME2_genes.txt', delim = '\t')

# ME13
ME13_df <- subset(module_df_7500, colors == "blue")
ME13_genes <- ME13_df$gene_id

me13 <- as.data.frame(ME13_genes)
write_delim(me13, '../../../tables/CD/train/WGCNA_Modules_Genes/ME13_genes.txt', delim = '\t')
```


```{r}
ora_ME0_df <- enrichGO(
  gene          = ME0_genes,
  universe = universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)

write_delim(ora_ME0_df[, c("ID", "Description")], "../../../tables/CD/train/WGCNA_Modules_Genes/pathways_ME0.txt", delim = "\t")
```

```{r}
ora_ME2_df <- enrichGO(
  gene          = ME2_genes,
  universe = universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)
write_delim(ora_ME2_df[, c("ID", "Description")], "../../../tables/CD/train/WGCNA_Modules_Genes/pathways_ME2.txt", delim = "\t")
```

```{r}
ora_ME13_df <- enrichGO(
  gene          = ME2_genes,
  universe = universe,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05
)
write_delim(ora_ME13_df[, c("ID", "Description")], "../../../tables/CD/train/WGCNA_Modules_Genes/pathways_ME13.txt", delim = "\t")
```


Data for Venn diagram
```{r}
downr_CD_4 <- rownames(data_plot_CD_DEG_results_4)[data_plot_CD_DEG_results_4$diffexpressed == "Downregulated"]
upr_CD_4 <- rownames(data_plot_CD_DEG_results_4)[data_plot_CD_DEG_results_4$diffexpressed == "Upregulated"]
```

#### Venn diagram
```{r}
gene_lists <- list(
  DEGs = rownames(CD_deg_sign_4_pval),
  ME0_genes = ME0_genes,
  ME2_genes = ME2_genes,
  ME13_genes = ME13_genes
)

# Build Venn diagram
venn.plot <- venn.diagram(
  x = gene_lists,
  filename = NULL, 
  fill = c("red", "blue", "green", "yellow"),
  alpha = 0.5,
  cat.col = c("red", "blue", "green", "black"),
  cat.cex = 1.2,
  margin = 0.1
)

# Вывод диаграммы
grid.newpage()
grid.draw(venn.plot)

png("../../../images/CD/train/venn_diagram.png", width = 2000, height = 2000, res = 300)
grid.draw(venn.plot)
dev.off()

```

Let's keep the genes common to tags (padj value < 0.05) and modules. Most of the genes overlap with module 13.
```{r}
intersect_values_ME13 <- intersect(rownames(CD_deg_sign_4_pval), ME13_genes)
intersect_values_ME2 <- intersect(rownames(CD_deg_sign_4_pval), ME2_genes)

```

```{r}
write.csv(intersect_values_ME13, file = "../../../tables/CD/train/intersect_genes_DEGs_and_ME13.csv", row.names = FALSE)

```
These genes were then analyzed using STRING and CytoHubba to search for hub-genes and further validate these genes in a validation cohort.






















































